tests/silent-payment.test.ts

```diff
@@ -240,3 +241,50 @@ it("can turn pubkey into taproot address", () => {
 it("can turn taproot address into pubkey", () => {
   assert.strictEqual(SilentPayment.addressToPubkey("bc1pgrhjjw52p6a03v635f7cnl6ttvuz9f34ujhaefm6xqtscd3m473szkl92g"), "40ef293a8a0ebaf8b351a27d89ff4b5b3822a635e4afdca77a30170c363bafa3");
 });
+
+it("can get pubkeys from Tx inputs", () => {
+  // txid 511e007f9c96b6d713a72b730506198f61dd96046edee72f0dc636bfe1f3a9cf
+  let tx = Transaction.fromHex(
+    "02000000000101e79e2690d05d3589257a5d1094de7f46bb1cfae3fc3fb3b644b790d4337931c5000000000001000000013226000000000000225120e92e6cb44492f87779999fbbc295540eef8a23f42efdebacac001ffa18074c100140692f4e81047496cd755c4a24b54ae36e74f7e303a265b1a9a643774d5699a6723cc66e9cdd395d2e487f7881a74bbb5740241498e70ede269583f862a3d47b4600000000"
+  );
+
+  const txPrevout0 = Transaction.fromHex('020000000001018f93f113a3d5f2d3feb7444ab8c8d7de5b2b2d1d9e9a5e2e2de42ad4b622958f0000000000000000800210270000000000002251203361aedbd209998e73f60ce0ea2245fa5c4ba747489c58eaa9222df401fda898ea140f000000000016001420c262ffbfe8be9744d502f421df8e1392f3231b02483045022100efaff08cc56bbe1a2819383ff95be23a6f6ab6acaf6bb75ccb7b1e462c62f1c202206fab1385f91ba4ae4fb7731944c8ebef501c7a16eafa320975cfb7697b8537ca012102fc490ee8b804b85d1b7d4959d0bd153ca5bc12fd82134aabbe96acb21b06a1ca00000000');
+
+  // important, need the locking script from previous transaction, and we put it in place of unlocking
+  // script so the util that parses pubkeys can find the pubkey
+  tx.ins[0].script = txPrevout0.outs[0].script;
```

Comment:
@junderw a lil guidance please? for this case to get pubkeys from inputs i have to assist it and provide locking script (input is p2tr). but for other testcases i can extract pubkeys from raw tx directly (inputs are p2wpkh). how can i tell by parsing txhex with bitcoinjs if i need to assist it and get prevout  tx..?

Reply:
You can look at the witness and see if it matches a known P2TR pattern.

The only 3 types that have witness are P2WPKH, P2WSH, and P2TR.

ONLY P2TR will have pubkey info in the previous output's script... so you need to figure out if the witness is the witness of a P2WSH, a P2WPKH, or P2TR.

You could try using the payments API and running the input script and input witness through a P2WPKH payment then a P2WSH payment and if both fail assume it's P2TR and needs extra data.